---
title: "Evaluations of KD aggregation methods"
author: "Fridolin Gross"
date: "`r Sys.Date()`"
output:
  rmdformats::readthedown:
    highlight: tango
    self_contained: true
    thumbnails: true
    lightbox: true
    gallery: false
    keep_md: no
    number_sections: yes
    fig_width: 8
    fig_height: 8
    fig_caption: true
    df_print: paged
    code_folding:	hide
---

# required packages:
```{r}
suppressPackageStartupMessages({
  library(ensembldb)
	library(AnnotationHub)
	library(BSgenome)
  library(org.Hs.eg.db)
  library(BSgenome.Hsapiens.UCSC.hg19)
  library(BSgenome.Hsapiens.UCSC.hg38)
  library(openxlsx)
  library(scanMiR)
  library(data.table)
  library(tidyverse)  
  })
```


### Prepare miRNA sequence file
```{r}
miRNA_seqs <- read.xlsx("../../Parameter_Optimization/Run_Bartel_Scripts/aav1741_DataS1.xlsx", rows=131:142, colNames = FALSE)

miRNA_seqs <- miRNA_seqs %>%
  mutate(mir = gsub("[\\(\\)]", "", X1)) %>%
  mutate(guide_seq = gsub("U", "T", X2)) %>%
  select(-X1, -X2) %>%
  mutate(guide_family = mir)

write.table(miRNA_seqs, "./inputs/mirseqs_hek.txt", quote=F, sep="\t", row.names = F)
```


### Using Bartel's annotation

```{r}
g <- rtracklayer::import.gff("../../Parameter_Optimization/data/McGeary_2019_raw/GSE140218_HEK293_transcriptannotations.gtf")
```

```{r}
gr_cds <- g[g$type=="CDS"]
gr_cds <- split(gr_cds,gr_cds$transcript_id)
gr_UTR <- g[g$type=="3UTR"]
gr_UTR <- split(gr_UTR, gr_UTR$transcript_id)

seq_cds <- extractTranscriptSeqs(BSgenome.Hsapiens.UCSC.hg19, gr_cds)
seq_UTR <- extractTranscriptSeqs(BSgenome.Hsapiens.UCSC.hg19, gr_UTR)

df_transcripts <- data.frame(
           transcript = names(seq_UTR),
           orf = as.character(seq_cds[names(seq_UTR)]),
           orf_length = width(seq_cds[names(seq_UTR)]),
           utr3 = as.character(seq_UTR),
           utr3_length = width(seq_UTR)
           )
df_transcripts$orf_utr3 <- paste(df_transcripts$orf, df_transcripts$utr3, sep="")
```


# get seqs
```{r}
names10 <- readRDS("../scanMiR/namesSeqs10.rds")
names100 <- readRDS("../scanMiR/namesSeqs100.rds")
names1000 <- readRDS("../scanMiR/namesSeqs1000.rds")
```


#10seqs 
```{r}
# subset
df_10 <- df_transcripts[names10,] 

# write table
write.table(df_10, "./inputs/hek_transcripts_10.txt", 
            sep="\t", quote=F, row.names = F)

# get sequences
orf_utr3_10 <- df_10$orf_utr3
names(orf_utr3_10) <- df_10$transcript
writeXStringSet(DNAStringSet(orf_utr3_10),"./inputs/hek_orf_utr3_10.fa")
```


#100seqs 
```{r}
# subset
df_100 <- df_transcripts[names100,] 

# write table
write.table(df_100, "./inputs/hek_transcripts_100.txt", 
            sep="\t", quote=F, row.names = F)

# get sequences
orf_utr3_100 <- df_100$orf_utr3
names(orf_utr3_100) <- df_100$transcript
writeXStringSet(DNAStringSet(orf_utr3_100),"./inputs/hek_orf_utr3_100.fa")
```



#1000seqs 
```{r}
# subset
df_1000 <- df_transcripts[names1000,] 

# write table
write.table(df_1000, "./inputs/hek_transcripts_1000.txt", 
            sep="\t", quote=F, row.names = F)

# get sequences
orf_utr3_1000 <- df_1000$orf_utr3
names(orf_utr3_1000) <- df_1000$transcript
writeXStringSet(DNAStringSet(orf_utr3_1000),"./inputs/hek_orf_utr3_1000.fa")
```
