---
title: "Eichhorn_analysis"
author: "Michael Soutschek"
date: "8 3 2021"
output: html_document
---


# Required packages:
```{r}
suppressPackageStartupMessages({
  library(tidyverse)
  library(SummarizedExperiment)
  library(GenomicRanges)
  library(edgeR)
  library(scanMiRData)
  library(ggsci)
  })
```



```{r}
get_r2 <- function(log_ag, m, dea, column, p3=0){
  m_agg <- scanMiR:::.aggregate_miRNA(m,ag = log_ag,p3 = p3,coef_orf = 0,coef_utr = 0)
  dea <- dea[c("Transcript.ID", column)]
  names(dea) <- c("Transcript.ID", "logFC")
  m_merged <- merge(m_agg, dea, by.x = "transcript", by.y = "Transcript.ID", all.y=TRUE)
  m_merged$repression[is.na(m_merged$repression)] <- 0
  return(cor(m_merged$repression, m_merged[["logFC"]])^2)
}

get_r2_vec <- function(m, log_ag_vec, dea, column, p3=0){
  r2 <- map_dbl(log_ag_vec, get_r2, m, dea, column, p3=p3)
  return(data.frame(log_ag = log_ag_vec, r2=r2))
}
```


# Load Data and generate SE
```{r}
if(file.exists("se_eichhorn.rds")){
  se_eichhorn <- readRDS("se_eichhorn.rds")
} else {
  dirs <- dir("/mnt/schratt/p1006/Eichhorn_2014/salmon2", full.names = FALSE)
  quant_files <- paste("/mnt/schratt/p1006/Eichhorn_2014/salmon2", dirs, "quant.sf", sep="/")
  names(quant_files) <- dirs
  quant_data <- lapply(quant_files, read.table, sep="\t", header = TRUE)
  counts_list <- lapply(quant_data, `[`, "NumReads")
  counts <- as.matrix(do.call(cbind, counts_list))
  colnames(counts) <- c("WT","122KO")
  TPM_list <- lapply(quant_data, `[`, "TPM")
  TPM <- as.matrix(do.call(cbind, TPM_list))
  SRA_Table_wt <- read.table("/mnt/schratt/p1006/Eichhorn_2014/SraRunTable_wt.txt", sep=",", header=TRUE)
   SRA_Table_ko <- read.table("/mnt/schratt/p1006/Eichhorn_2014/SraRunTable_ko.txt", sep=",", header=TRUE)
   SRA_Table <- rbind(SRA_Table_wt,SRA_Table_ko)
  cd <- data.frame(condition = rep(c("WT", "122KO"), each=1), row.names = SRA_Table$Run[c(2,4)])
  cd$condition <- factor(cd$condition, levels = c("WT", "122KO"))
  rd <- quant_data[[1]][c("Length", "EffectiveLength")]
  row.names(rd) <- quant_data[[1]][["Name"]]
  se_eichhorn <- SummarizedExperiment(list(counts = counts, TPM = TPM), colData=cd, rowData = rd)
  saveRDS(se_eichhorn, file='se_eichhorn.rds')
}
```



# DEA
```{r}
if(file.exists("se_eichhorn_DEA.rds")){
  se_eichhorn_DEA <- readRDS("se_eichhorn_DEA.rds")
} else {
  se_eichhorn_DEA <- se_eichhorn[filterByExpr(assay(se_eichhorn),group = cd$condition),]
  dds <- DGEList(assays(se_eichhorn_DEA)$counts, group = cd$condition)
  dds <- calcNormFactors(dds)
  dds <- estimateGLMCommonDisp(dds,method="deviance",robust="TRUE",subset=NULL)
  res <- as.data.frame(exactTest(dds))
  rowData(se_eichhorn_DEA)$logFC <- res[rownames(se_eichhorn_DEA),]$logFC
  rowData(se_eichhorn_DEA)$logCPM <- res[rownames(se_eichhorn_DEA),]$logCPM
  colData(se_eichhorn_DEA)$lib_size <- assay(se_eichhorn_DEA) %>% colSums()
  assays(se_eichhorn_DEA)$logCPM <- log1p(cpm(dds, normalized.lib.sizes = T))
  saveRDS(se_eichhorn_DEA, file='se_eichhorn_DEA.rds')
}
```

# Read in and pre-process data
```{r}
## DEA results
se_eichhorn_DEA <- readRDS("se_eichhorn_DEA.rds")
eichhorn_logFC <- as.data.frame(rowData(se_eichhorn_DEA)["logFC"])
eichhorn_logFC$TPM <- rowMeans(assays(se_eichhorn_DEA)[["TPM"]])
eichhorn_logFC$logCPM <- rowData(se_eichhorn_DEA)[["logCPM"]]
eichhorn_logFC$Transcript.ID <- gsub("\\..*", "", row.names(eichhorn_logFC))
```


## scan
```{r}
#scan results
if(file.exists("mmu_full_matches_122_5p.rds")){
  m <- readRDS("mmu_full_matches_122_5p.rds")
} else {
  mods <- getKdModels(species = "mmu",NULL)
  mods_122 <- mods[["mmu-miR-122-5p"]]
  m <- runFullScan(species = "mmu", mods = mods_122,UTRonly = FALSE,cores = 12,maxLogKd = c(-0.3,-0.3),save.path = FALSE)
  saveRDS(m,"mmu_full_matches_122_5p.rds")
}

#aggregation results
if(file.exists("mmu_agg_full_matches_122_5p.rds")){
  m_agg <- readRDS("mmu_agg_full_matches_122_5p.rds")
} else {
  m_agg <- scanMiR:::.aggregate_miRNA(m,p3 = 0L,coef_orf = 0,coef_utr = 0)
  saveRDS(m_agg,"mmu_agg_full_matches_122_5p.rds")
}
```


# Get Targetscan results
```{r}
if(file.exists("Targetscan_all_122_5p.rds")){
  TS_122_m <- readRDS("Targetscan_all_122_5p.rds")
} else {
  TS <- fread("../Amin_clean/Summary_Counts.txt")
  TS$sites <- TS[["Total num conserved sites"]]+TS[["Total num nonconserved sites"]]
  fam <- fread("../Amin_clean/miR_Family_Info.txt")
  fam_122 <- fam[fam$`MiRBase ID` == "mmu-miR-122-5p",]
  TS_122 <- TS[TS$`miRNA family` == fam_122$`Seed+m8`,]
  TS_122_m <- TS_122[grep("mmu-",TS_122$`Representative miRNA`),]
  TS_122_m$score <- TS_122_m$`Total context++ score`
  TS_122_m$Transcript.ID <- gsub("\\..*","",TS_122_m$`Transcript ID`)
  saveRDS(TS_122_m,"Targetscan_all_122_5p.rds")
}


if(file.exists("Targetscan_Gene_Info.rds")){
  TS_gene_info <- readRDS("Targetscan_Gene_Info.rds")
} else {
  tmp <- tempfile()
  download.file(
          "http://www.targetscan.org/mmu_72/mmu_72_data_download/Gene_info.txt.zip", tmp)
  TS_gene_info <- fread(unzip(file.path(tmp)))
  TS_gene_info <- as.data.frame(TS_gene_info)
  TS_gene_info <- TS_gene_info[TS_gene_info$`Representative transcript?` == "1",]
  TS_gene_info <- TS_gene_info[TS_gene_info$`3P-seq tags` >= 5,]
  dupl <- TS_gene_info[which(duplicated(TS_gene_info$`Gene symbol`)),"Gene symbol"]
  dupl_ex <- character()
  for(i in dupl){
    df <- TS_gene_info[TS_gene_info$`Gene symbol` == i,]
    df <- df[order(df$`3P-seq tags`,decreasing = TRUE),]
    dupl_ex[i] <- df[1,"Transcript ID"]
  }
  TS_add <- TS_gene_info[TS_gene_info$`Transcript ID` %in% dupl_ex,]
  TS_red <- TS_gene_info[!TS_gene_info$`Gene symbol` %in% dupl,]
  TS_info <- rbind(TS_red,TS_add)
  TS_info$`Transcript.ID` <- gsub("\\..*","",TS_info$`Transcript ID`)
  TS_gene_info <- TS_info
  saveRDS(TS_gene_info,"Targetscan_Gene_Info.rds")
}
```




# Filter the Data Set
```{r}
tx2gene <- read.delim("/mnt/reference/reference/Mus_musculus/Ensembl/GRCm38/Annotation/Release_99-2020-10-proteinCoding/tx2gene", col.names = c("Transcript.ID", "Gene.ID", "Gene.Name"), header = FALSE, sep='|') %>%
  mutate(Gene.Name = gsub("\t.*", "", Gene.Name))

eichhorn_logFC_filt_tx <- merge(eichhorn_logFC, tx2gene, by="Transcript.ID") %>%
  as.data.frame() %>%
  mutate(CPM = 2^logCPM) %>%
  group_by(Gene.ID) %>%
  mutate(fracCPM = CPM/sum(CPM)) %>%
  ungroup() %>%
  dplyr::filter(fracCPM > 0.9) %>%
  dplyr::filter(TPM > 0.5) %>%
  dplyr::filter(Transcript.ID %in% TS_gene_info$Transcript.ID)
```



# Correlation analysis
```{r}
if(file.exists("plot_df.rds")){
  df <- readRDS("plot_df.rds")
} else {
log_ag_vec <- seq(-8, 0, 0.5)
mir = "mmu-miR-122-5p"


get_r2_df <- function(m, dea,mir, tool,type="all", p3=0, level="gene") {
  r2 <- get_r2_vec(m, log_ag_vec, dea, "logFC", p3=p3)
  r2 <- cbind(r2, type = type,mir = mir,tool = tool, level=level, p3=p3)
  return(r2)
}

df <- get_r2_df(m,eichhorn_logFC_filt_tx,mir = mir,tool = "scanMiR",type = "tx_filter", level = "transcript")

# Get Targetscan Corr.
TS_122_m <- as.data.frame(TS_122_m)
merged_TS_tx <- merge(eichhorn_logFC_filt_tx[c("Transcript.ID", "logFC")], TS_122_m[c("Transcript.ID", "score")], by="Transcript.ID", all.x = TRUE)
merged_TS_tx$score[is.na(merged_TS_tx$score)] <- 0
merged_TS_tx$score[merged_TS_tx$score == "NULL"] <- 0
merged_TS_tx$score <- as.numeric(merged_TS_tx$score)
r2_TS_tx <- cor(merged_TS_tx[["logFC"]], merged_TS_tx[["score"]])^2
df_TS_tx <- data.frame(log_ag = log_ag_vec, r2=r2_TS_tx, tool="TS", type = "tx_filter", level = "transcript", p3=0,mir = mir)
df <- rbind(df, df_TS_tx)
saveRDS(df,"plot_df.rds")
}


# Potentially add p3 and length
# test <- get_r2_df(m,p3 = 0.044,amin_logFC_filt_tx,mir = mir,tool = "scanMiR",type = "tx_filter", level = "transcript")
```



```{r}
ggplot(df, aes(x=log_ag, y=r2, color=tool)) +
  geom_line() +
  theme_bw() + 
  scale_color_nejm() +
  ggtitle("miR-122 ko (Eichhorn et al., 2014)")

ggsave("Eichhorn.png", width=6, height=5)
```










