---
title: "Figure5_TDMD"
author: "Pierre-Luc Germain"
date: "4/7/2021"
output: html_document
---

```{r}
suppressPackageStartupMessages({
  library(rtracklayer)
  library(GenomicRanges)
  library(ggplot2)
  library(ggrepel)
  library(cowplot)
  library(readxl)
  library(SummarizedExperiment)
  library(scanMiR)
  library(scanMiRApp)
  library(fst)
})
theme_set(theme_cowplot())
```


#colors
```{r}
source("../misc.R")
```


#Get miRNA expression and TDMD miRs from Shi et al
```{r}
mirexp_df <- read.delim("../../Add_Datasets/Chiang/Normalized miRNA Brain expression from Chiang et al. 2010_.csv", sep=",")
colnames(mirexp_df) <- c("miRNA","cpm")
mirexp <- sapply(split(mirexp_df[,2], mirexp_df[,1]), na.rm=TRUE, FUN=max)

#load tdmd mirs
mir_tdmd <- read_excel("../../Add_Datasets/Shi/Shi_et_al_Supp_abc9359-Data-S2.xlsx", 
    sheet = "D. Induced mouse neurons", skip = 1)
mir_tdmd <- mir_tdmd[!is.na(mir_tdmd$p_adj),]
mir_tdmd_sig <- mir_tdmd[mir_tdmd$p_adj < 10e-5,"miRNA", drop = TRUE]
mouse_mir_tdmd_sig <- paste0("mmu-",mir_tdmd_sig)
```


#Get expressed transcripts from Whipple et al.
```{r}
se <- readRDS("../../Add_Datasets/Whipple_2020/se_whipple.rds")
se_wt <- se[,se$condition == "WT"]
se_wt <- se_wt[which(rowSums(assays(se_wt)$TPM >= 10) >= 2),]
se_wt_transcripts <- gsub("\\..*","",row.names(se_wt))
se_wt_TPM <- as.data.frame(assays(se_wt)$TPM)
se_wt_TPM$mean_TPM <- rowMeans(se_wt_TPM)
se_wt_TPM$seqnames <- gsub("\\..*","",row.names(se_wt_TPM))
```




#load mRNA scan and filter for TDMD and expressed transcripts

```{r, eval=FALSE}
m <- loadIndexedFst("/mnt/schratt/miRNA_KD/pl_test/new/mm10.matches.fst")
# filter for TDMD
tmp <-fst(m@fst.file)
note <- tmp[,"note"]
w <- grep("TDMD", note)
tmp_tdmd <- tmp[w,]
# filter for transcripts
transcripts <- tmp_tdmd[,"seqnames"]
w2 <- transcripts %in% se_wt_transcripts
tmp_tdmd_tr <- tmp_tdmd[w2,]
# filter for tdmd miRs
tmp_tdmd_tr$mir <- gsub("mmu-","",tmp_tdmd_tr$miRNA)
tdmd <- tmp_tdmd_tr[tmp_tdmd_tr$mir %in% mir_tdmd_sig,]
tdmd$biotype <- "mRNA"
tdmd$mir <- NULL
saveRDS(tdmd,"mouse_tdmd_expr_transcripts_mirs.rds")
```


```{r, include=FALSE}
tdmd <- readRDS("mouse_tdmd_expr_transcripts_mirs.rds")
```


#load lncRNA scan and filter for expressed transcripts
```{r}
tdmd_linc <- readRDS("../../lncRNA_TDMD/mouse_lncRNAs_tdmd_miRNAs_TDMD.rds")
tdmd_linc_expr <- tdmd_linc[seqnames(tdmd_linc) %in% se_wt_transcripts]
tdmd_linc_expr$biotype <- "lncRNA"
tdmd_linc_expr <- as.data.frame(tdmd_linc_expr)
tdmd_linc_expr <- tdmd_linc_expr[,!colnames(tdmd_linc_expr) %in%  c("end","strand","width","p3.mir.bulge","p3.target.bulge","p3.mismatch")]
tdmd_linc_expr$ORF <- "FALSE"
```



#Combine
```{r}
tdmd_combined <- rbind(tdmd_linc_expr,tdmd)
tdmd_combined <- merge(tdmd_combined,se_wt_TPM[,c("seqnames","mean_TPM")],by = "seqnames")
tdmd_combined$miRNA <- gsub("mmu-","",tdmd_combined$miRNA)
tdmd_combined <- merge(tdmd_combined,mir_tdmd,by = "miRNA")
tdmd_combined$label <- paste(tdmd_combined$seqnames, tdmd_combined$miRNA, sep="\n")
saveRDS(tdmd_combined,"tdmd.rds")
```

```{r}
tdmd_combined <- readRDS("tdmd.rds")
tdmd_combined$biotype <- factor(as.character(tdmd_combined$biotype),levels = c("mRNA","lncRNA"))
```


#Plot
```{r}
w <- (tdmd_combined$mean_TPM > 100 & tdmd_combined$`norm WT` > 500 & tdmd_combined$ORF == "FALSE") | tdmd_combined$label == "ENSMUST00000153581\nmiR-7a-5p"
p1 <- ggplot(tdmd_combined, aes(x=mean_TPM,y = log10(`norm WT`), colour=ORF, shape=biotype, label=label)) + 
    geom_point(size=2) + scale_x_sqrt() + ylim(1,4) + 
    labs(x="Transcript expression (TPM)", y="normalized miRNA expression (log10)", shape="Biotype") + 
    geom_text_repel(data=tdmd_combined[w,], min.segment.length = 0, show.legend = FALSE,nudge_x = 1, nudge_y = c(-0.25,0.5,0.35,-0.5,0.35)) +
    scale_color_manual(values = c("darkred",darkturquoise)) +
    scale_fill_manual(values = c("darkred",darkturquoise))
p1
```






```{r}
g <- import("../../circRNAs/Mouse/mouse_brain_circRNAs.gtf.gz")
tmp <- strsplit(g$isoform,"|",fixed=TRUE)
g <- g[rep(seq_along(g),lengths(tmp))]
g$isoform <- unlist(tmp)
df <- data.frame(row.names=paste(g$circ_id, g$isoform), score=as.integer(g$score), 
                 gene=g$gene_name, strand=as.factor(strand(g)))
df <- df[order(df$gene, -df$score),]

circ <- readRDS("../../circRNAs/Mouse/canonical.GR.rds")
circ <- circ[grep("TDMD",circ$note)]
circ$tx.type <- "circRNA"
circ$tx.exp <- df[as.character(seqnames(circ)),"score"]
circ$gene <- df[as.character(seqnames(circ)),"gene"]
circ$mir.exp <- log10(1+mirexp[as.character(circ$miRNA)])
circ <- circ[circ$miRNA %in% mouse_mir_tdmd_sig]
circ$gene.label <- sapply(strsplit(circ$gene,","),"[",1)
circ.table <- as.data.frame(circ)[,-c(3:5)]
saveRDS(circ.table, "suppTable_circTDMD.rds")
circ$label <- paste(circ$gene.label, circ$miRNA, sep="\n")
circ1 <- as.data.frame(circ[which(circ$mir.exp > 0 & circ$tx.exp > 0),])
circ2 <- as.data.frame(circ[which(circ$mir.exp > log10(51) & circ$tx.exp > 10),])
circ1 <- circ1[order(circ1$type, -circ1$tx.exp),]
circ1 <- circ1[!duplicated(circ1[,c("gene.label","miRNA")]),]
```



```{r}
w <- which(circ1$tx.exp>300 & circ1$mir.exp > 0.5)
circ1$label <- gsub("mmu-","",circ1$label)
circ1$lab.dir <- sign(circ1$mir.exp - (5.5-0.06*sqrt(circ1$tx.exp)))
p2 <- ggplot(circ1, aes(tx.exp, mir.exp, shape=type, label=label)) + 
  geom_point(size=3,color = "#332288") + scale_x_sqrt() + 
  labs(x="circRNA expression in the brain (BSJ counts)", y="miRNA expression (logCPMs)", shape="Match\ntype") + 
  geom_text_repel(data=circ1[w,], nudge_x = 5, nudge_y = 0, min.segment.length = 0, show.legend = FALSE,force_pull = -0.05)
p2
```





#grid
```{r, fig.width=8, fig.height=10}
pp <- plot_grid( 
  p1 + ggtitle("Potential TDMD sites in induced mouse neurons"),
  p2 + ggtitle("Potential TDMD sites in circRNAs"), 
  nrow=2, labels="AUTO", scale=0.95
)
ggsave("Figure_5.pdf", plot = pp, width = 8, height = 8, bg="white")
ggsave("Figure_5.png", plot = pp, width = 8, height = 8, bg="white")
```











