---
title: "Predicted miRNA repression in real dataset"
author: "Fridolin Gross"
date: "`r Sys.Date()`"
output:
  rmdformats::readthedown:
    highlight: tango
    self_contained: true
    thumbnails: true
    lightbox: true
    gallery: false
    keep_md: no
    number_sections: yes
    fig_width: 8
    fig_height: 8
    fig_caption: true
    df_print: paged
    code_folding:	hide
---

# Required packages:
```{r}
suppressPackageStartupMessages({
  library(tidyverse)
  library(SummarizedExperiment)
  library(GenomicRanges)
  library(edgeR)
  library(scanMiRData)
  library(ggsci)
  })

```



```{r}
get_r2 <- function(log_ag, m, dea, column, p3=0){
  m_agg <- scanMiR:::.aggregate_miRNA(m,a = log_ag,p3 = p3,coef_orf = 0,coef_utr = 0)
  dea <- dea[c("Transcript.ID", column)]
  names(dea) <- c("Transcript.ID", "logFC")
  m_merged <- merge(m_agg, dea, by.x = "transcript", by.y = "Transcript.ID", all.y=TRUE)
  m_merged$repression[is.na(m_merged$repression)] <- 0
  return(cor(m_merged$repression, m_merged[["logFC"]])^2)
}

get_r2_vec <- function(m, log_ag_vec, dea, column, p3=0){
  r2 <- map_dbl(log_ag_vec, get_r2, m, dea, column, p3=p3)
  return(data.frame(log_ag = log_ag_vec, r2=r2))
}
```



# Load Data and generate SE
```{r}
if(file.exists("se_amin.rds")){
  se_amin <- readRDS("se_amin.rds")
} else {
  dirs <- dir("./salmon", full.names = FALSE)
  quant_files <- paste("./salmon", dirs, "quant.sf", sep="/")
  names(quant_files) <- dirs
  quant_data <- lapply(quant_files, read.table, sep="\t", header = TRUE)
  counts_list <- lapply(quant_data, `[`, "NumReads")
  counts <- as.matrix(do.call(cbind, counts_list))
  TPM_list <- lapply(quant_data, `[`, "TPM")
  TPM <- as.matrix(do.call(cbind, TPM_list))
  SRA_Table <- read.table("/mnt/schratt/p1006/Amin_2015/SraRunTable.txt", sep=",", header=TRUE)
  cd <- data.frame(condition = rep(c("WT", "218DKO"), each=2), row.names = dirs)
  cd$condition <- factor(cd$condition, levels = c("WT", "218DKO"))
  rd <- quant_data[[1]][c("Length", "EffectiveLength")]
  row.names(rd) <- quant_data[[1]][["Name"]]
  se_amin <- SummarizedExperiment(list(counts = counts, TPM = TPM), colData=cd, rowData = rd)
  saveRDS(se_amin, file='se_amin.rds')
}
```

# DEA
```{r}
if(file.exists("se_amin_DEA.rds")){
  se_amin_DEA <- readRDS("se_amin_DEA.rds")
} else {
  mm <- model.matrix(~condition, data=cd)
  se_amin_DEA <- se_amin[filterByExpr(assay(se_amin), mm),]
  
  dds <- DGEList(assays(se_amin_DEA)$counts, group=cd$condition)
  dds <- calcNormFactors(dds)
  dds <- estimateDisp(dds, mm)
  fit <- glmFit(dds, mm)
  res <- as.data.frame(topTags(glmLRT(fit, 'condition218DKO'), Inf))
  
  rowData(se_amin_DEA)$logFC <- res[rownames(se_amin_DEA),]$logFC
  rowData(se_amin_DEA)$logCPM <- res[rownames(se_amin_DEA),]$logCPM
  rowData(se_amin_DEA)$FDR <- res[rownames(se_amin_DEA),]$FDR
  rowData(se_amin_DEA)$logFC <- res[rownames(se_amin_DEA),]$logFC
  
  colData(se_amin_DEA)$lib_size <- assay(se_amin_DEA) %>% colSums()
  assays(se_amin_DEA)$logCPM <- log1p(cpm(dds, normalized.lib.sizes = T))
  
  saveRDS(se_amin_DEA, file='se_amin_DEA.rds')
}
```


# Read in and pre-process data
```{r}
## DEA results
se_amin_DEA <- readRDS("se_amin_DEA.rds")
amin_logFC <- as.data.frame(rowData(se_amin_DEA)["logFC"])
amin_logFC$TPM <- rowMeans(assays(se_amin_DEA)[["TPM"]])
amin_logFC$logCPM <- rowData(se_amin_DEA)[["logCPM"]]
amin_logFC$Transcript.ID <- gsub("\\..*", "", row.names(amin_logFC))
```

## scan
```{r}
#scan results
if(file.exists("mmu_full_matches_218_5p.rds")){
  m <- readRDS("mmu_full_matches_218_5p.rds")
} else {
  mods <- getKdModels(species = "mmu",NULL)
  mods_218 <- mods[["mmu-miR-218-5p"]]
  m <- runFullScan(species = "mmu", mods = mods_218,UTRonly = FALSE,cores = 12,maxLogKd = c(-0.3,-0.3),save.path = FALSE)
  saveRDS(m,"mmu_full_matches_218_5p.rds")
}

#aggregation results
if(file.exists("mmu_agg_full_matches_218_5p.rds")){
  m_agg <- readRDS("mmu_agg_full_matches_218_5p.rds")
} else {
  m_agg <- scanMiR:::.aggregate_miRNA(m,p3 = 0L,coef_orf = 0,coef_utr = 0)
  saveRDS(m_agg,"mmu_agg_full_matches_218_5p.rds")
}
```




# Get Targetscan8 results
```{r}
if(file.exists("Targetscan8_all_218_5p.rds")){
  TS_218_m <- readRDS("Targetscan8_all_218_5p.rds")
} else {
  tmp <- tempfile()
  download.file(
          "http://www.targetscan.org/mmu_80/mmu_80_data_download/Summary_Counts.all_predictions.txt.zip", tmp)
  TS <- fread(unzip(file.path(tmp)),drop = c("Aggregate PCT"))
  TS$sites <- TS[["Total num conserved sites"]]+TS[["Total num nonconserved sites"]]
  fam <- enrichMiR:::.getTargetscan_miRfamilies(species = "mouse")
  fam_218 <- fam[fam$MiRBase.ID == "mmu-miR-218-5p",]
  TS_218 <- TS[TS$`miRNA family` == fam_218$Seed.m8,]
  TS_218_m <- TS_218[grep("mmu-",TS_218$`Representative miRNA`),]
  TS_218_m$score <- TS_218_m$`Total context++ score`
  TS_218_m$occ_high <- TS_218_m$`Predicted occupancy - high miRNA`
  TS_218_m$occ_low <- TS_218_m$`Predicted occupancy - low miRNA`
  TS_218_m$occ_trans <- TS_218_m$`Predicted occupancy - transfected miRNA`
  TS_218_m$Transcript.ID <- gsub("\\..*","",TS_218_m$`Transcript ID`)
  saveRDS(TS_218_m,"Targetscan8_all_218_5p.rds")
}


if(file.exists("Targetscan8_Gene_Info.rds")){
  TS_gene_info <- readRDS("Targetscan8_Gene_Info.rds")
} else {
  tmp <- tempfile()
  download.file(
          "http://www.targetscan.org/mmu_80/mmu_80_data_download/Gene_info.txt.zip", tmp)
  TS_gene_info <- fread(unzip(file.path(tmp)))
  TS_gene_info <- as.data.frame(TS_gene_info)
  TS_gene_info <- TS_gene_info[TS_gene_info$`Representative transcript?` == "1",]
  TS_gene_info <- TS_gene_info[TS_gene_info$`3P-seq tags` >= 5,]
  dupl <- TS_gene_info[which(duplicated(TS_gene_info$`Gene symbol`)),"Gene symbol"]
  dupl_ex <- character()
  for(i in dupl){
    df <- TS_gene_info[TS_gene_info$`Gene symbol` == i,]
    df <- df[order(df$`3P-seq tags`,decreasing = TRUE),]
    dupl_ex[i] <- df[1,"Transcript ID"]
  }
  TS_add <- TS_gene_info[TS_gene_info$`Transcript ID` %in% dupl_ex,]
  TS_red <- TS_gene_info[!TS_gene_info$`Gene symbol` %in% dupl,]
  TS_info <- rbind(TS_red,TS_add)
  TS_info$`Transcript.ID` <- gsub("\\..*","",TS_info$`Transcript ID`)
  TS_gene_info <- TS_info
  saveRDS(TS_gene_info,"Targetscan8_Gene_Info.rds")
}
```





# Get Targetscan7 results
```{r}
# if(file.exists("Targetscan_all_218_5p.rds")){
#   TS_218_m <- readRDS("Targetscan_all_218_5p.rds")
# } else {
#   tmp <- tempfile()
#   download.file(
#           "http://www.targetscan.org/mmu_72/mmu_72_data_download/Summary_Counts.all_predictions.txt.zip", tmp)
#   TS <- fread(unzip(file.path(tmp)),drop = c("Aggregate PCT"))
#   TS$sites <- TS[["Total num conserved sites"]]+TS[["Total num nonconserved sites"]]
#   fam <- enrichMiR:::.fetch_Mirfamilies(species = "mouse")
#   fam_218 <- fam[fam$MiRBase.ID == "mmu-miR-218-5p",]
#   TS_218 <- TS[TS$`miRNA family` == fam_218$Seed.m8,]
#   TS_218_m <- TS_218[grep("mmu-",TS_218$`Representative miRNA`),]
#   TS_218_m$score <- TS_218_m$`Total context++ score`
#   TS_218_m$Transcript.ID <- gsub("\\..*","",TS_218_m$`Transcript ID`)
#   saveRDS(TS_218_m,"Targetscan_all_218_5p.rds")
# }
# 
# 
# if(file.exists("Targetscan_Gene_Info.rds")){
#   TS_gene_info <- readRDS("Targetscan_Gene_Info.rds")
# } else {
#   tmp <- tempfile()
#   download.file(
#           "http://www.targetscan.org/mmu_72/mmu_72_data_download/Gene_info.txt.zip", tmp)
#   TS_gene_info <- fread(unzip(file.path(tmp)))
#   TS_gene_info <- as.data.frame(TS_gene_info)
#   TS_gene_info <- TS_gene_info[TS_gene_info$`Representative transcript?` == "1",]
#   TS_gene_info <- TS_gene_info[TS_gene_info$`3P-seq tags` >= 5,]
#   dupl <- TS_gene_info[which(duplicated(TS_gene_info$`Gene symbol`)),"Gene symbol"]
#   dupl_ex <- character()
#   for(i in dupl){
#     df <- TS_gene_info[TS_gene_info$`Gene symbol` == i,]
#     df <- df[order(df$`3P-seq tags`,decreasing = TRUE),]
#     dupl_ex[i] <- df[1,"Transcript ID"]
#   }
#   TS_add <- TS_gene_info[TS_gene_info$`Transcript ID` %in% dupl_ex,]
#   TS_red <- TS_gene_info[!TS_gene_info$`Gene symbol` %in% dupl,]
#   TS_info <- rbind(TS_red,TS_add)
#   TS_info$`Transcript.ID` <- gsub("\\..*","",TS_info$`Transcript ID`)
#   TS_gene_info <- TS_info
#   saveRDS(TS_gene_info,"Targetscan_Gene_Info.rds")
# }
```




# Filter the Amin Set
```{r}
tx2gene <- read.delim("/mnt/reference/reference/Mus_musculus/Ensembl/GRCm38/Annotation/Release_99-2020-10-proteinCoding/tx2gene", col.names = c("Transcript.ID", "Gene.ID", "Gene.Name"), header = FALSE, sep='|') %>%
  mutate(Gene.Name = gsub("\t.*", "", Gene.Name))

amin_logFC_filt_tx <- merge(amin_logFC, tx2gene, by="Transcript.ID") %>%
  as.data.frame() %>%
  mutate(CPM = 2^logCPM) %>%
  group_by(Gene.ID) %>%
  mutate(fracCPM = CPM/sum(CPM)) %>%
  ungroup() %>%
  dplyr::filter(fracCPM > 0.9) %>%
  dplyr::filter(TPM > 10) %>%
  dplyr::filter(Transcript.ID %in% TS_gene_info$Transcript.ID)


```


# Correlation analysis
```{r}
if(file.exists("plot_df.rds")){
  df <- readRDS("plot_df.rds")
} else {
log_ag_vec <- exp(seq(-8, 0, 0.5))
mir = "mmu-miR-218-5p"


get_r2_df <- function(m, dea,mir, tool,type="all", p3=0, level="gene") {
  r2 <- get_r2_vec(m, log_ag_vec, dea, "logFC", p3=p3)
  r2 <- cbind(r2, type = type,mir = mir,tool = tool, level=level, p3=p3)
  return(r2)
}

df <- get_r2_df(m,amin_logFC_filt_tx,mir = mir,tool = "scanMiR",type = "tx_filter", level = "transcript")

# Get Targetscan Corr.
TS_218_m <- as.data.frame(TS_218_m)
merged_TS_tx <- merge(amin_logFC_filt_tx[c("Transcript.ID", "logFC")], TS_218_m[c("Transcript.ID", "score","occ_high","occ_low","occ_trans")], by="Transcript.ID", all.x = TRUE)
merged_TS_tx[is.na(merged_TS_tx)] <- 0
merged_TS_tx$score[merged_TS_tx$score == "NULL"] <- 0
merged_TS_tx$occ_high[merged_TS_tx$occ_high == "NULL"] <- 0
merged_TS_tx$occ_low[merged_TS_tx$occ_low == "NULL"] <- 0
merged_TS_tx$occ_trans[merged_TS_tx$occ_trans == "NULL"] <- 0

merged_TS_tx$score <- as.numeric(merged_TS_tx$score)
merged_TS_tx$occ_high <- as.numeric(merged_TS_tx$occ_high)
merged_TS_tx$occ_low <- as.numeric(merged_TS_tx$occ_low)
merged_TS_tx$occ_trans <- as.numeric(merged_TS_tx$occ_trans)

r2_TS7_tx <- cor(merged_TS_tx[["logFC"]], merged_TS_tx[["score"]])^2
r2_TS8_high_tx <- cor(merged_TS_tx[["logFC"]], merged_TS_tx[["occ_high"]])^2
r2_TS8_low_tx <- cor(merged_TS_tx[["logFC"]], merged_TS_tx[["occ_low"]])^2
r2_TS8_trans_tx <- cor(merged_TS_tx[["logFC"]], merged_TS_tx[["occ_trans"]])^2

df_TS7_tx <- data.frame(log_ag = log_ag_vec, r2=r2_TS7_tx, tool="TS7", type = "tx_filter", level = "transcript", p3=0,mir = mir)
df_TS8_high_tx <- data.frame(log_ag = log_ag_vec, r2=r2_TS8_high_tx, tool="TS8_high", type = "tx_filter", level = "transcript", p3=0,mir = mir)
# df_TS8_low_tx <- data.frame(log_ag = log_ag_vec, r2=r2_TS8_low_tx, tool="TS8_low", type = "tx_filter", level = "transcript", p3=0,mir = mir)
# df_TS8_trans_tx <- data.frame(log_ag = log_ag_vec, r2=r2_TS8_trans_tx, tool="TS8_trans", type = "tx_filter", level = "transcript", p3=0,mir = mir)

df <- rbind(df, df_TS7_tx,df_TS8_high_tx)
saveRDS(df,"plot_df.rds")
}


# Potentially add p3 and length
# test <- get_r2_df(m,p3 = 0.044,amin_logFC_filt_tx,mir = mir,tool = "scanMiR",type = "tx_filter", level = "transcript")
```

```{r}
ggplot(df, aes(x=log(log_ag), y=r2, color=tool)) +
  geom_line() +
  theme_bw() + 
  scale_color_nejm() +
  ggtitle("miR-218 ko (Amin et al., 2015)")

ggsave("Amin.png", width=6, height=5)
```

